File: 0567-permutation-in-string/0567-permutation-in-string.py
from collections import defaultdict
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n = len(s1)
        m = len(s2)
        if n > m: return False
        fq1 = defaultdict(int)
        fq2 = defaultdict(int)
        for s in s1:
            fq1[s] += 1
        for i in range(n):
            fq2[s2[i]] += 1 
        if fq1 == fq2: return True    
        i = 0
        j = n 
        while j < m:
            fq2[s2[j]] += 1
            fq2[s2[i]] -= 1
            if fq2[s2[i]] == 0: del fq2[s2[i]]
            i += 1
            j += 1
            if fq1 == fq2: return True
        return False               



                    
        

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n = s1.size();
        int m = s2.size();
        if(m < n)return false;
        vector<int> fqS1(26, 0);
        vector<int> fqS2(26, 0);
        for(int i = 0; i < n; i++){
            fqS1[s1[i] - 'a']++;
            fqS2[s2[i] - 'a']++;
        }
        if(fqS1 == fqS2) return true;
        for(int i = n; i < m; i++){
            fqS2[s2[i] - 'a']++;
            fqS2[s2[i-n] - 'a']--;
            if(fqS1 == fqS2) return true;
        }
        return false;
    }
};